### PR TITLE
#254 patch create and add article to issue map endpoint

### PR DESCRIPTION
**Pull request by shuerry**
## Pull Request

[comment]: <> (Leave blank if a section does not apply.)

### Brief Summary
[comment]: <> (Put a brief summary of your changes.)
- Created create and add article to issue map endpoint

### Questions / Considerations for the Future
[comment]: <> (Note any questions, or things to note that might involve this PR in the future.)
- 
- 

### API Changes
[comment]: <> (Note any new endpoints, deleted endpoints, or updated endpoints.)
- New routes for create and add article
- New function in UserAccessor to fetch all the users with the given emails
- Fixed the getArticle method in ArticleAccessor

### Database Changes
[comment]: <> (Note changes to any database schemas.)
- 
- 

### New Tests
[comment]: <> (Note any new or modified test cases.)
- PatchIssueMap tests
- 

[comment]: <> (Put the ticket # this PR closes.)
Closes # 254

### CODE REVIEW

File: app/controllers/issueMapController.js [hunk 1]
```diff
@@ -1,7 +1,133 @@
...
+   * method to create and add an article from the issue map.
+   *
+   * Create and add an article from the issue map.
+   *
+   * @param {Request} req
+   * @param {Response} res
+   */
+  static async addAndCreateArticle(req, res) {
+    try {
+      console.log("Request Body:", req.body);
```
Comment (suttonspindler) [inline app/controllers/issueMapController.js:36]: Console logs are useful for debugging - just be sure to remove them (except from test files) before you commit your code

File: app/controllers/issueMapController.js [hunk 2]
```diff
@@ -1,7 +1,133 @@
...
+      console.log("Article Created:", createdArticle);
+
+      const issueMap = await IssueMapAccessor.getIssueMapByIssueNumber(issueNumber);
+
+      if (!issueMap) {
+        console.error(`Issue Map for issueNumber ${issueNumber} not found.`);
+        throw new ErrorIssueMapNotFound();
+      }
+
+      console.log("Fetched Issue Map:", issueMap);
```
Comment (suttonspindler) [inline app/controllers/issueMapController.js:104]: Remove console log

File: app/controllers/issueMapController.js [hunk 3]
```diff
@@ -1,7 +1,133 @@
...
+        issueMap.articles.push(createdArticle._id);
+      }
+
+      issueMap.modificationTime = new Date();
+      await issueMap.save();
+
+      console.log("Issue Map Updated and Saved");
+      return res.status(200).json(issueMap);
+    } catch (e) {
+      console.error("Error Occurred:", e.message);
```
Comment (suttonspindler) [inline app/controllers/issueMapController.js:125]: Redundant printing of error message

File: app/controllers/issueMapController.js [hunk 4]
```diff
@@ -1,7 +1,133 @@
...
+
+      console.log("Fetched Issue Map:", issueMap);
+
+      if (section) {
+        const sectionIndex = issueMap.sections.findIndex((sec) => sec.sectionName === section);
+
+        if (sectionIndex >= 0) {
+          issueMap.sections[sectionIndex].articles.push(createdArticle._id);
+        } else {
+          console.error("Section Not Found");
```
Comment (suttonspindler) [inline app/controllers/issueMapController.js:112]: Remove console.error, throwing ErrorSectionNotFound does same thing

File: app/controllers/issueMapController.js [hunk 5]
```diff
@@ -1,7 +1,133 @@
...
+        modificationTime: new Date(),
+      };
+
+      const createdArticle = await Article.create(newArticle);
+      console.log("Article Created:", createdArticle);
+
+      const issueMap = await IssueMapAccessor.getIssueMapByIssueNumber(issueNumber);
+
+      if (!issueMap) {
+        console.error(`Issue Map for issueNumber ${issueNumber} not found.`);
```
Comment (suttonspindler) [inline app/controllers/issueMapController.js:100]: Remove console.error, just use ErrorIssueMapNotFound to handle the error

File: app/controllers/issueMapController.js [hunk 6]
```diff
@@ -1,7 +1,133 @@
...
+        photographers: photographerUsers,
+        ArticleContent: [],
+        comments: [],
+        sources: [],
+        creationTime: new Date(),
+        modificationTime: new Date(),
+      };
+
+      const createdArticle = await Article.create(newArticle);
+      console.log("Article Created:", createdArticle);
```
Comment (suttonspindler) [inline app/controllers/issueMapController.js:95]: remove console log

File: tests/__tests__/api/issue_map/patchIssueMap.js [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+      .set("Cookie", [`token=${tokens["ethan@ethan.com"]}`])
+      .send(requestBody);
+
+    console.log(response.body);
+    expect(response.status).toBe(200);
+    const createdArticleID = response.body.articles.at(-1);
+    const article = await ArticlesAccessor.getArticle(createdArticleID);
+    console.log(article);
+    expect(article.issueNumber).toStrictEqual(requestBody.issueNumber);
+    expect(article.slug).toStrictEqual(requestBody.articleSlug);
```
Comment (suttonspindler) [inline tests/__tests__/api/issue_map/patchIssueMap.js:177]: Consider adding a few more .toStrictEqual checks here to confirm that designStatus is set to Needs_Designer, photographyStatus is set to Needs_Photographer, etc

File: tests/__tests__/api/issue_map/patchIssueMap.js [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    const sectionIndex = response.body.sections.findIndex((sec) => sec.sectionName === validSectionName);
+
+    const createdArticleID = response.body.sections[sectionIndex].articles.at(-1);
+    console.log(response.body.sections[sectionIndex]);
+
+    const article = await ArticlesAccessor.getArticle(createdArticleID);
+    console.log(article);
+
+    expect(article.issueNumber).toStrictEqual(requestBody.issueNumber);
+    expect(article.slug).toStrictEqual(requestBody.articleSlug);
```
Comment (suttonspindler) [inline tests/__tests__/api/issue_map/patchIssueMap.js:210]: Same here - consider adding a few more .toStrictEqual checks here to confirm that designStatus is set to Has_Designer, photographyStatus is set to Has_Photographer, etc

File: app/controllers/issueMapController.js [hunk 9]
```diff
@@ -1,7 +1,123 @@
...
+      const existingArticle = await ArticlesAccessor.getArticleBySlug(articleSlug);
+      if (issueNumber < 0 || !articleSlug || pageLength < 0 || existingArticle) {
+        throw new ErrorInvalidRequestBody();
+      }
+
+      const articleStatus = ArticleStatus.Print;
+      const designStatus = designers.length > 0 ? DesignStatus.Has_Designer : DesignStatus.Needs_Designer;
+      const photographyStatus =
+        photographers.length > 0 ? PhotographyStatus.Photographer_Assigned : PhotographyStatus.Needs_Photographer;
+      const writingStatus = editors.length > 0 ? WritingStatus.Has_Editor : WritingStatus.Needs_Editor;
```
Comment (ethanszeto) [inline app/controllers/issueMapController.js:56]: Great

**Pull request merged**